opt: exec builder support for Limit/Offset

RaduBerinde opened the pull request:
#### opt: fix limit build issue

We weren't handling the case when the select is behind parens.

Release note: None

#### opt: exec builder support for Limit/Offset

Also renaming `build` directive to `opt` in the execbuilder test, for
consistency with other tests.

Release note: None

#### opt: fix child physical properties of limit

We were accidentally requiring no properties of the Limit/Offset input
if there is no requirement on the Limit/Offset itself.

Release note: None

RaduBerinde committed:
opt: exec builder support for Limit/Offset

Also renaming `build` directive to `opt` in the execbuilder test, for
consistency with other tests.

Release note: None

RaduBerinde committed:
opt: fix limit build issue

We weren't handling the case when the select is behind parens.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/23756)
<!-- Reviewable:end -->

RaduBerinde committed:
opt: fix child physical properties of limit

We were accidentally requiring no properties of the Limit/Offset input
if there is no requirement on the Limit/Offset itself.

Release note: None

RaduBerinde commented:
Added a commit with another fix.

RaduBerinde commented:
Note: I am working on a `FilterUnusedLimitCols` rule, I will send that out in a subsequent PR.

andy-kimball commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 12 files reviewed at latest revision, all discussions resolved, all commit checks successful.

---

*[pkg/sql/opt/xform/physical_props_factory.go, line 131 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23756#-L7Rcz8693GrWKIVxMW1:-L7Rcz8693GrWKIVxMW2:b-rg1kfc) ([raw file](https://github.com/cockroachdb/cockroach/blob/26d74cd0c46cfa17d248c0aef591db2c55de7098/pkg/sql/opt/xform/physical_props_factory.go#L131)):*
> ```Go
> func (f physicalPropsFactory) constructChildProps(ev ExprView, nth int) opt.PhysicalPropsID {
> 	if ev.required == opt.MinPhysPropsID {
> 		switch ev.Operator() {
> ```

This code makes me wonder if we should be propagating the required ordering above the Limit in `buildLimit` in `optbuilder`. If I'm reading the current `buildLimit` code correctly, we are propagating the ordering, which may cause unnecessary sorts in more complex cases. While we require the `Limit` input to be sorted, we don't require the `Limit` output to be sorted. This doesn't affect this PR, but we should remember to address it elsewhere.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23756#-:-L7ReDkDAc8Pya_XqJzL:bnfp4nl)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: 0 of 12 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/sql/opt/xform/physical_props_factory.go, line 131 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23756#-L7Rcz8693GrWKIVxMW1:-L7SEuaD2Y8r-iq3durf:btrfoui) ([raw file](https://github.com/cockroachdb/cockroach/blob/26d74cd0c46cfa17d248c0aef591db2c55de7098/pkg/sql/opt/xform/physical_props_factory.go#L131)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

This code makes me wonder if we should be propagating the required ordering above the Limit in `buildLimit` in `optbuilder`. If I'm reading the current `buildLimit` code correctly, we are propagating the ordering, which may cause unnecessary sorts in more complex cases. While we require the `Limit` input to be sorted, we don't require the `Limit` output to be sorted. This doesn't affect this PR, but we should remember to address it elsewhere.
</blockquote></details>

A query like `SELECT * FROM t ORDER BY x LIMIT 10` does require the final output to be sorted. Maybe I'm not understanding your point, do you have an example?

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23756)*
<!-- Sent from Reviewable.io -->

andy-kimball commented:
Review status: 0 of 12 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/sql/opt/xform/physical_props_factory.go, line 131 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23756#-L7Rcz8693GrWKIVxMW1:-L7SS_yJ3ilRpg5AESet:ba62976) ([raw file](https://github.com/cockroachdb/cockroach/blob/26d74cd0c46cfa17d248c0aef591db2c55de7098/pkg/sql/opt/xform/physical_props_factory.go#L131)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

A query like `SELECT * FROM t ORDER BY x LIMIT 10` does require the final output to be sorted. Maybe I'm not understanding your point, do you have an example?
</blockquote></details>

You're right in general, I was thinking of cases where a query like that is nested within a larger query:

`SELECT * FROM (SELECT * FROM t ORDER BY x LIMIT 10) a, b WHERE a.x=b.x`

We don't want an extra sort to happen around the `a` subquery. But I think the existing code would work correctly in this case (i.e. not trigger an extra sort).

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23756)*
<!-- Sent from Reviewable.io -->
